File: golang/264.ugly-number-ii.go
package main

import (
	"container/heap"
)

/*
 * [264] Ugly Number II
 *
 * https://leetcode.com/problems/ugly-number-ii/description/
 *
 * algorithms
 * Medium (33.58%)
 * Total Accepted:    78.1K
 * Total Submissions: 232.4K
 * Testcase Example:  '10'
 *
 * Write a program to find the n-th ugly number.
 *
 * Ugly numbers are positive numbers whose prime factors only include 2, 3,
 * 5.
 *
 * Example:
 *
 *
 * Input: n = 10
 * Output: 12
 * Explanation: 1, 2, 3, 4, 5, 6, 8, 9, 10, 12 is the sequence of the first 10
 * ugly numbers.
 *
 * Note:
 *
 *
 * 1 is typically treated as an ugly number.
 * n does not exceed 1690.
 *
 *
 */
func nthUglyNumber(n int) int {
	if n <= 0 {
		return -1
	}

	h := &IntHeap{Data: []int{1}}
	cur := 0

	x := 0
	for cur < n {
		x = heap.Pop(h).(int)
		cur++

		if x%3 > 0 && x%5 > 0 {
			heap.Push(h, x*2)
			heap.Push(h, x*3)
		} else if x%5 > 0 {
			heap.Push(h, x*3)
		}

		heap.Push(h, x*5)

	}

	return x
}

type IntHeap struct {
	Data []int
}

func (h IntHeap) Len() int {
	return len(h.Data)
}

func (h IntHeap) Less(i int, j int) bool {
	return h.Data[i] < h.Data[j]
}

func (h IntHeap) Swap(i int, j int) {
	h.Data[i], h.Data[j] = h.Data[j], h.Data[i]
}

func (h *IntHeap) Push(x interface{}) {
	h.Data = append(h.Data, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	N := h.Len()
	res := h.Data[N-1]
	h.Data = h.Data[:N-1]
	return res
}

File: cpp/264.ugly-number-ii.cpp
/*
 * @lc app=leetcode.cn id=264 lang=cpp
 *
 * [264] 丑数 II
 *
 * https://leetcode.cn/problems/ugly-number-ii/description/
 *
 * algorithms
 * Medium (58.43%)
 * Total Accepted:    165.6K
 * Total Submissions: 283.6K
 * Testcase Example:  '10'
 *
 * 给你一个整数 n ，请你找出并返回第 n 个 丑数 。
 * 
 * 丑数 就是质因子只包含 2、3 和 5 的正整数。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：n = 10
 * 输出：12
 * 解释：[1, 2, 3, 4, 5, 6, 8, 9, 10, 12] 是由前 10 个丑数组成的序列。
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：n = 1
 * 输出：1
 * 解释：1 通常被视为丑数。
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 1 <= n <= 1690
 * 
 * 
 */
class Solution {
public:
    int nthUglyNumber(int n) {
      std::vector<int> dp(n+1);
      dp[1] = 1;

      int p2 = 1;
      int p3 = 1;
      int p5 = 1;
      for (int i = 2; i <= n; i++) {
        int v2 = dp[p2] * 2;
        int v3 = dp[p3] * 3;
        int v5 = dp[p5] * 5;
        dp[i] = min({v2, v3, v5});
        if (dp[i] == v2) {
          p2++;
        }

        if (dp[i] == v3) {
          p3++;
        }

        if (dp[i] == v5) {
          p5++;
        }
      }

      return dp[n];
    }

    int nthUglyNumber1(int n) {
      if (n <= 0) {
        return 0;
      }

      std::priority_queue<long, std::vector<long>, std::greater<long>> que;
      std::unordered_set<int> visited{1};

      que.push(1);

      for (int i = 0; i < n-1; i++) {
        int cur = que.top();
        que.pop();
        std::vector<long> values = {2l * cur, 3l * cur, 5l * cur};
        for (auto val : values) {
          if (!visited.contains(val)) {
            que.push(val);
            visited.insert(val);
          }
        }
      }

      return que.top();
    }
};
